js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -493,6 +509,217 @@ function checkReservedFieldNames(fields): void {
         }
     });
 }
+
+export type FieldPath = string | string[];
+
+/**
+ * Get assigned field paths for the record.
+ *
+ * See test cases for example of what this looks like
+ */
+export function getAssignedFieldsDeep(record: ViewModelInterface<any, any>): FieldPath[] {
+    const fieldNames: FieldPath[] = [];
+    for (const fieldName of record._assignedFields as string[]) {
+        const data = record[fieldName];
+        if (isViewModelInstance(data)) {
+            for (const path of getAssignedFieldsDeep(data)) {
+                fieldNames.push([fieldName, ...(Array.isArray(path) ? path : [path])]);
+            }
+        } else {
+            fieldNames.push(fieldName);
+        }
+    }
+    return fieldNames;
+}
+
+/**
+ * Flatten a nested path to a single level with dot notation
+ *
+ * eg.
+ * ```
+ * flattenFieldPath([
+ *   'id',
+ *   ['user', ['group', 'id']],
+ *   ['user', 'groupId'],
+ *   ['user', 'id'],
+ *   'userId'
+ * ])
+ * // ['id', 'user.group.id', 'user.groupId', 'user.id', 'userId']
+ * ```
+ */
+export function flattenFieldPath(fieldPath: FieldPath[] | FieldPath, separator = '.'): string[] {
+    if (typeof fieldPath === 'string') {
+        return [fieldPath];
+    }
+    const flattenedPath: string[] = [];
+    for (const path of fieldPath) {
+        if (typeof path === 'string') {
+            flattenedPath.push(path);
+        } else {
+            flattenedPath.push(path.join(separator));
+        }
+    }
+    return flattenedPath;
+}
+
+export function expandField(model: ViewModelConstructor<any>, fieldName: string): FieldPath[] {
+    const field = model.fields[fieldName];
+    if (field instanceof RelatedViewModelField) {
+        return field.to.fieldNames
+            .filter(
+                subFieldName => !(field.to.fields[subFieldName] instanceof RelatedViewModelField)
+            )
+            .map(subFieldName => [fieldName, subFieldName]);
+    }
+    return [fieldName];
+}
+
+/**
+ * Thrown when attempting to access a field that does not exist on a ViewModel
+ */
+export class InvalidFieldError extends Error {}
+
+/**
+ * For the given field paths expand any relation fields to include the nested non-relation fields
```

Comment:
fields on a related model yes - but not recursive (hence the non-relation fields only)

Reply:
Can you give an example so that's clearer

Reply:
The expanded explanation drops the mention of the fact that it doesn't include non-relation fields (prob want to update the examples to include one field, eg something along the lines of "note that the group's countryId field is included but not the country field because country is a RelatedViewModelField and wasn't specifically mentioned")

Reply:
I have a commit not yet pushed that says:

> We exclude [RelatedViewModelField](doc:RelatedViewModelField)'s to avoid circular dependencies. In the example > above this means the `owner` field is excluded but the source field `ownerId` is still included.

(also note it's an internal presto function, not exported for public consumption)